File: largest_color_value_in_a_directed_graph.py
class Solution:
    def largestPathValue(self, colors: str, edges: List[List[int]]) -> int:
        graph = defaultdict(list)
        # indegree of vertex is the number of edges going to the vertex
        indegree = [0] * len(colors)

        for edge in edges:
            if edge[0] == edge[1]:
                return -1
            graph[edge[0]].append(edge[1])
            indegree[edge[1]] += 1
        
        q = deque([i for i in range(len(colors)) if indegree[i] == 0])
        dp = [[0] * 26 for _ in range(len(colors))]
        max_color = 0

        while q:
            vertex = q.popleft()
            color = ord(colors[vertex]) - ord('a')
            dp[vertex][color] += 1
            max_color = max(max_color, dp[vertex][color])

            for nei in graph[vertex]:
                indegree[nei] -= 1
            
                for c in range(26):
                    dp[nei][c] = max(dp[nei][c], dp[vertex][c])
                
                if indegree[nei] == 0:
                    q.append(nei)

        if any(indegree):
            return -1

        return max_color

File: largest_color_value_in_a_directed_graph.js
/**
 * @param {string} colors
 * @param {number[][]} edges
 * @return {number}
 */
var largestPathValue = function(colors, edges) {
    const graph = new Map();
    const indegree = new Array(colors.length).fill(0);

    for (const [u, v] of edges) {
        if (u === v) return -1;
        graph.set(u, (graph.get(u) || []).concat(v));
        indegree[v]++;
    }

    const q = [];
    for (let i = 0; i < colors.length; i++) {
        if (indegree[i] === 0) q.push(i);
    }

    const dp = Array.from({ length: colors.length }, () => Array(26).fill(0));
    let maxColor = 0;

    while (q.length) {
        const vertex = q.shift();
        const color = colors.charCodeAt(vertex) - 97;
        dp[vertex][color]++;
        maxColor = Math.max(maxColor, dp[vertex][color]);

        if (!graph.has(vertex)) continue;

        for (const nei of graph.get(vertex)) {
            indegree[nei]--;

            for (let c = 0; c < 26; c++) {
                dp[nei][c] = Math.max(dp[nei][c], dp[vertex][c]);
            }

            if (indegree[nei] === 0) q.push(nei);
        }
    }

    return indegree.some(x => x > 0) ? -1 : maxColor;    
};
